### DISCUSSION TITLE
Add `py:` prefix as namespace for our own PyScript packages

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#2061] [Proposals]
AFAIK in micropip or pip in general, packages with a colon in the name [are not allowed](https://packaging.python.org/en/latest/specifications/name-normalization/), meaning if this is the RegExp for a package name `^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$` we should be good by deciding that `py:web`, as example, as *config* `packages = [...]` entry should never collide or interfere with other packages in the wild.

There is previous work in the *JS* world too, see `node:fs` VS `bun:fs` or other "*protocols*" used to disambiguate the final package to use, hence I am pretty confident this is not a terrible idea.

### Why

Our core is growing and growing but not everyone will need `pyscript.pyweb`, as example, among others, but we do have the possibility to pre-parse the config and instrument it so that:

  * packages that contain `py:` prefix will be removed from the packages list that will use the native underlying way to import packages on bootstrap
  * those packages will instead point at a `zip` or `tar.gz` folder that contains everything they need to work and we sneakily move and orchestrate those packages through the `files` entry in the config, augmenting it to bootstrap via unzip/untargz operations the dependency
  * we might then somehow relate those modules as `pyscript` exports but (*imho*) we'd be also good to have `pyweb` as result so that one simply `import pyweb` when `py:web` package is defined
  * users that want to use our extra modules can simply add a config entry and we do the rest
  * this would work with the `ltk:core` or `ltk:thing` namespace too, it's up to us how many prefixes we want to enable, as long as those modules are properly archived and available via CDN too (or for offline usage as well)

I think the only thing to solve, if everyone agrees on this way forward, is to decide how we can also make those packages related *JS* modules available, but I think we could simply decide ourselves that those packages will have a list of modules defined as `[js_modules]` (let's say that's the optional `js_modules.toml` file that should be present at the root of the unzipped package) so that we can also instrument that field in the resulting config and everything *should* play out as smoothly as possible and as transparent (for users) as possible.

The convention for ourselves is then:

  * create a folder, let's say a `pyweb` folder
  * put `__init__.py` in there and every other file that's needed
  * optionally put a `js_modules.toml` (or `js_modules.json`) at the root of that folder
  * the config dance will filter/ignore `py:` packages from the `package` field and will instead automatically extract those folders in the virtual FS
  * if `js_modules.*` is found, those modules augment the already available `js_modules` entry in the config or create one if not there yet
  * users just add `packages = ["py:web"]` in their config and they are good to go

@fpliger, @ntoll, @JeffersGlass, @mchilvers, what are your thoughts around this proposal?

**Comment by ntoll** [comment]
@laffra (given mention of ltk) may also have opinions.

If I remember, @fpliger wasn't 100% certain the names `pyweb` and `pydom` were set in stone. But I like this proposal as a first step in modularising these (and other) aspects of our namespaces out of the core asset.

**Reply by WebReflection** [reply]
the name is not an issue, the convention is ... I think assuming `py:dom` results into a `import pydom` is the key/core thing we need to agree about, if that's `py:element` or `py:stuff` it doesn't matter.

My gut says that landing these as `pyscript` extra modules *might* be the best we can do, as it's less obtrusive (let's say a *pydom* package already exist on *pip* we're going to have conflicts) but I don't know if prefixing with `py` any *pip* module is actually a common thing or an absolute "*nobody does that*" in practice.

**Comment by WebReflection** [comment]
**amend** the `js_modules.toml` or `js_modules.json` is a bad idea as it confines the scope to js modules only ... that's the only concern it'd solve but at the same time it's future hostile so that we might want to enforce a `config.toml` or `config.json` at the root of the package so that it's easier to reason about and no extra documentation would be needed ... it's just the package config.

That, in turns, allow `pyweb` or `pydom` to eventually define its own `packages` as dependency so that we can concatenate that within the already user's defined `config` and seamlessly integrate with the current *config* logic.
